### PR TITLE
Add Snowflake REAL type to C type conversions

### PR DESCRIPTION
**Pull request by sfc-gh-pbulawa**
### TL;DR

Added support for Snowflake's REAL logical type (FLOAT, DOUBLE, REAL) in the ODBC driver with proper type conversions and comprehensive test coverage.

### Changes

- **Added REAL type conversion support**: Implemented `SnowflakeReal` converter in `odbc/src/conversion/real.rs` that handles conversions from Float64Array to various ODBC C data types including SQL_C_DOUBLE (default), SQL_C_FLOAT, integer types, SQL_C_BIT, SQL_C_CHAR, SQL_C_WCHAR, SQL_C_NUMERIC, and SQL_C_BINARY

- **Integrated REAL type into conversion framework**: Added `Real(real::SnowflakeReal)` variant to `SnowflakeFieldType` enum and registered "REAL" logical type mapping in the conversion system

- **Added comprehensive test coverage**: Created extensive unit tests in `real_tests.rs` covering all conversion scenarios, edge cases, overflow conditions, and SQLSTATE compliance (01S07 for fractional truncation, 22003 for numeric overflow)

- **Added integration tests**: Implemented full ODBC integration tests in `real_tests.cpp` testing default conversions, explicit type conversions, NULL handling, buffer management, and multi-row scenarios

- **Updated behavior differences documentation**: Added entries BD#16 and BD#17 documenting differences from the old driver regarding SQL_C_BINARY conversion (now returns SQL_NUMERIC_STRUCT instead of raw bytes) and SQL_C_CHAR buffer truncation handling (now returns SQL_SUCCESS_WITH_INFO instead of SQL_ERROR)

- **Enhanced test infrastructure**: Added `Float64Array` import and `field_with_real_meta()` helper function to support REAL type testing in the data conversion test suite

The implementation follows ODBC specification requirements for type conversions, proper error handling with appropriate SQLSTATEs, and maintains compatibility with existing Snowflake FLOAT/DOUBLE/REAL column types.

### CODE REVIEW

**Comment by sfc-gh-pbulawa** [comment]
* **#367** <a href="https://app.graphite.com/github/pr/snowflakedb/universal-driver/367?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#305** <a href="https://app.graphite.com/github/pr/snowflakedb/universal-driver/305?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/snowflakedb/universal-driver/305?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#303** <a href="https://app.graphite.com/github/pr/snowflakedb/universal-driver/303?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 2 other dependent PRs ([#304](https://github.com/snowflakedb/universal-driver/pull/304) <a href="https://app.graphite.com/github/pr/snowflakedb/universal-driver/304?utm_source=stack-comment-icon" target="_blank">Graphite</a>, [#318](https://github.com/snowflakedb/universal-driver/pull/318) <a href="https://app.graphite.com/github/pr/snowflakedb/universal-driver/318?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: odbc/src/conversion/real.rs [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+                        }
+                        .fail();
+                    }
+                }
+                Ok(warnings)
+            }
+            CDataType::Binary => {
+                let int_value = snowflake_value.trunc() as i64;
+                let abs_value = (int_value.unsigned_abs() as u128).to_le_bytes();
+                let sign: u8 = if int_value >= 0 { 1 } else { 0 };
```
[the remark this answers is not in the document]
Comment (sfc-gh-pbulawa) [reply odbc/src/conversion/real.rs]: Done

File: odbc/src/conversion/real.rs [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+    let truncated = value.trunc();
+    if truncated < min || truncated > max {
+        NumericValueOutOfRangeSnafu {
+            reason: format!("Value {value} is out of range ({min} to {max})"),
+        }
+        .fail()
+    } else {
+        Ok(())
+    }
+}
```
[the remark this answers is not in the document]
Comment (sfc-gh-makowalski) [reply odbc/src/conversion/real.rs:51]: \+1

File: odbc/src/conversion/real.rs [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+                    }
+                    .fail();
+                }
+
+                let numeric = sql::Numeric {
+                    precision: binding.precision.unwrap_or(38) as u8,
+                    scale: target_scale as i8,
+                    sign: if is_negative { 0 } else { 1 },
+                    val: (int_scaled as u128).to_le_bytes(),
+                };
```
[the remark this answers is not in the document]
Comment (sfc-gh-makowalski) [reply odbc/src/conversion/real.rs:195]: \+1

File: odbc/src/conversion/real.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (sfc-gh-makowalski) [reply odbc/src/conversion/real.rs]: did you push the change?

**Review by sfc-gh-makowalski — CHANGES_REQUESTED** [review]
Overall the PR looks good. I bumped some of the copilot comments. I think we should add more tests around NaN and infinity, negative fractional -> unsigned, negative zero and **SQL_C_NUMERIC with negative values and non-zero scale.  
  
I'm** **willing** **to** **approve** **when** **these** **tests** **are** **provided**

File: odbc_tests/tests/datatype_tests/real_tests.cpp [hunk 4]
```diff
@@ -0,0 +1,1109 @@
+
```
Comment (sfc-gh-jszczerbinski) [inline odbc_tests/tests/datatype_tests/real_tests.cpp:1]: In general let's move those to e2e and provide gherkins, same for number tests

**Review by sfc-gh-jszczerbinski — CHANGES_REQUESTED** [review]

File: odbc_tests/tests/datatype_tests/real_tests.cpp [hunk 5]
```diff
@@ -0,0 +1,1350 @@
...
+      "SELECT 2.2250738585072014e-308::DOUBLE, "
+      "       1e-307::DOUBLE, "
+      "       -2.2250738585072014e-308::DOUBLE");
+
+  double v1 = get_data_default_as_double(stmt, 1);
+  double v2 = get_data_default_as_double(stmt, 2);
+  double v3 = get_data_default_as_double(stmt, 3);
+
+  CHECK(v1 > 0.0);
+  CHECK(v1 < 1e-300);
```
Comment (sfc-gh-jszczerbinski) [inline odbc_tests/tests/datatype_tests/real_tests.cpp:129]: Use catch macros for float comparision

File: odbc_tests/tests/datatype_tests/real_tests.cpp [hunk 6]
```diff
@@ -0,0 +1,1350 @@
...
+// ODBC Spec: SQLSTATE 01S07 — Fractional truncation warning
+// Converting a FLOAT value with fractional digits to an integer C type
+// should return SQL_SUCCESS_WITH_INFO with SQLSTATE 01S07.
+// ============================================================================
+
+TEST_CASE("REAL fractional truncation returns 01S07", "[datatype][real][01S07]") {
+  Connection conn;
+  auto random_schema = Schema::use_random_schema(conn);
+
+  SECTION("SQL_C_LONG") {
```
Comment (sfc-gh-jszczerbinski) [inline odbc_tests/tests/datatype_tests/real_tests.cpp]: Don't use sections for section perf

File: odbc_tests/BehaviorDifferences.yaml [hunk 7]
```diff
@@ -108,4 +108,71 @@ behavior_differences:
       Impact: Applications that set descriptor fields for SQL_C_NUMERIC will
       now see the correct precision and scale in the output struct, matching
       the ODBC specification behavior.
-      TODO: To be fixed in old driver
\ No newline at end of file
+      TODO: To be fixed in old driver
+
+  16:
+    name: "SQL_C_BINARY on FLOAT/DOUBLE/REAL columns returns raw f64 bytes instead of SQL_NUMERIC_STRUCT"
```
Comment (sfc-gh-jszczerbinski) [inline odbc_tests/BehaviorDifferences.yaml:114]: Let's write test to see what SQL type SQLDescribeCol returns. Propably this is expected driver behavior

File: odbc_tests/BehaviorDifferences.yaml [hunk 8]
```diff
@@ -108,4 +108,71 @@ behavior_differences:
...
+      any integer or BIT type.
+      Note: Both drivers correctly reject NaN for SQL_C_NUMERIC (returning 22003) and
+      both correctly produce the string "NaN" for SQL_C_CHAR. The discrepancy is limited
+      to integer and BIT targets.
+      Impact: Applications that previously received 0 for NaN inputs to integer types
+      will now receive an error. This is a spec-compliance improvement.
+      TODO: To be fixed in old driver
+
+  19:
+    name: "SQL_C_NUMERIC produces negative zero for FLOAT values like -0.5 that truncate to zero magnitude"
```
Comment (sfc-gh-jszczerbinski) [inline odbc_tests/BehaviorDifferences.yaml:165]: I think this is expected behavior

**Review by sfc-gh-jszczerbinski — APPROVED** [review]

**Review by sfc-gh-makowalski — APPROVED** [review]
